IP: Add API to subscribe/unsubscribe to multicast addresses.

LuDuda opened the pull request:
This PR fixes #891 issue by adding external API to subscribe and unsubscribe multicast addresses.
Also i have added possibility of setting the multicast promiscuous mode suggested by @nibanks to accept all, where for example we act as an NCP.

There were no external registration for multicast addresses, so i mostly reused same code as it is for unicast addresses.

I follow numenclature of Linux and `ip maddr` command, so the CLI command is called `ipmaddr`.

The only problematic case is with command that list multicast addresses. Currently none of the default "ff02::1, ff03::1 etc." are printed, since they are not in the list of multicast addresses. Those addresses are just checked in `Netif::IsMulticastSubscribed()` method which saves a lot of RAM. So i would vote for leaving it as it is.

jwhui commented on include/openthread.h:
```diff
@@ -791,6 +791,72 @@ ThreadError otAddUnicastAddress(otInstance *aInstance, const otNetifAddress *aAd
+ * @sa otIsMulticastPromiscuousModeEnabled
+ * @sa otEnableMulticastPromiscuousMode
+ */
+void otDisableMulticastPromiscuousMode(otInstance *aInstance);
```
As pointed out by @nibanks, should we add a `bool` argument to `otEnableMulticastPromiscuousMode()` and remove this function?

pvanhorn replied:
In openthread.h we have API's of each flavor;  otDisableMacBlacklist, otAddMacWhitelist one the one hand and otSetRouterRoleEnabled on the other.

In general, it would be nice if our API's were consistently one type or the other, but given that we have multiple instances of both types already in existence, there is not much point in trying to enforce a rule now.

pvanhorn replied:
My point is that when you build consensus, you conclude with a rule that you apply moving forward.  Given the lack of consistency in our existing API's, a meaningful consensus can't logically be reached, unless someone is willing to go back and update all the existing API's which violate the agreed to consensus/rule.

Since the majority of our existing bool setting API's (Add/Remove, Enable/Disable) appear to be implemented with a pair of complimentary API's as we have here, I would prefer that we leave this as is.  My choice is strictly based on what I see as being the dominant style.

LuDuda replied:
Okay, then i will this PR as it is. If you need help with refactoring this please let me know.

jwhui commented on src/core/net/netif.cpp:
```diff
@@ -122,8 +123,12 @@ bool Netif::IsMulticastSubscribed(const Address &aAddress) const
 
-    if (aAddress.IsLinkLocalAllNodesMulticast() || aAddress.IsRealmLocalAllNodesMulticast() ||
-        aAddress.IsRealmLocalAllMplForwarders())
+    if (mMulticastPromiscuousMode)
```
I think we only want to apply the promiscuous mode to messages passed up to the host via `otReceiveIp6DatagramCallback`.  Applying the logic here will disable any multicast group filtering applied to Thread-internal multicast messages.

jwhui commented on src/core/net/netif.cpp:
```diff
@@ -200,6 +210,102 @@ ThreadError Netif::UnsubscribeMulticast(const NetifMulticastAddress &aAddress)
+    }
+
+    // Make sure we haven't set all the bits in the mask already
+    VerifyOrExit(mMaskExtUnicastAddresses != ((1 << OPENTHREAD_CONFIG_MAX_EXT_MULTICAST_IP_ADDRS) - 1),
```
`mMaskExtUnicastAddresses ` -> `mMaskExtMulticastAddresses`

LuDuda replied:
Ahh.. Good catch, will update all this copy paste things 👍

jwhui commented on src/core/net/netif.cpp:
```diff
@@ -200,6 +210,102 @@ ThreadError Netif::UnsubscribeMulticast(const NetifMulticastAddress &aAddress)
+
+    if (aAddressIndexToRemove != -1)
+    {
+        mMaskExtUnicastAddresses &= ~(1 << aAddressIndexToRemove);
```
`mMaskExtUnicastAddresses ` -> `mMaskExtMulticastAddresses`

jwhui commented on src/core/net/netif.cpp:
```diff
@@ -200,6 +210,102 @@ ThreadError Netif::UnsubscribeMulticast(const NetifMulticastAddress &aAddress)
+    assert(index < OPENTHREAD_CONFIG_MAX_EXT_MULTICAST_IP_ADDRS);
+
+    // Increase the count and mask the index
+    mMaskExtUnicastAddresses |= 1 << index;
```
`mMaskExtUnicastAddresses ` -> `mMaskExtMulticastAddresses`

jwhui commented on src/core/net/netif.cpp:
```diff
@@ -200,6 +210,102 @@ ThreadError Netif::UnsubscribeMulticast(const NetifMulticastAddress &aAddress)
+                 error = kThreadError_NoBufs);
+
+    // Get next available entry index
+    while ((mMaskExtUnicastAddresses & (1 << index)) != 0)
```
`mMaskExtUnicastAddresses ` -> `mMaskExtMulticastAddresses`

LuDuda commented:
I rebased and fixed mask variable. Also i applied the promiscuous mode to messages passed up to the host via otReceiveIp6DatagramCallback only as you requested.
